### PR TITLE
add tolerance for empty choices

### LINKED ISSUE #4213: `finish_reason` not set in AzureOpenAIChatCompletionClient.create_stream
### What happened?

The provided code snippet works fine for the `.create` call of `AzureOpenAIChatCompletionClient`, but errors on `.create_stream` call:

```
Creating client with config: {'model': 'gpt-4o', 'azure_endpoint': 'https://xxxxxxxx.openai.azure.com', 'azure_deployment': 'gpt-4o', 'api_version': '2024-08-01-preview', 'model_capabilities': {'vision': False, 'function_calling': False, 'json_output': False}, 'azure_ad_token_provider': <function get_bearer_token_provider.<locals>.wrapper at 0x108205da0>}
-----> Print output of .create call
/Users/mohammadmazraeh/Projects/autogen/python/packages/autogen-core/samples/distributed-group-chat/test_aoi.py:26: UserWarning: Resolved model mismatch: gpt-4o-2024-08-06 != gpt-4o-2024-05-13. Model mapping may be incorrect.
  single_output = await client.create(messages=messages)
-----> CreateResult(finish_reason='stop', content='The autumn leaves painted the park in vibrant shades of red and gold.', usage=RequestUsage(prompt_tokens=17, completion_tokens=14), cached=False, logprobs=None) - * 50
-----> Print output of .create_stream call
Traceback (most recent call last):
  File "/Users/mohammadmazraeh/Projects/autogen/python/packages/autogen-core/samples/distributed-group-chat/test_aoi.py", line 34, in <module>
    asyncio.run(main())
  File "/usr/local/Cellar/python@3.11/3.11.10/Frameworks/Python.framework/Versions/3.11/lib/python3.11/asyncio/runners.py", line 190, in run
    return runner.run(main)
           ^^^^^^^^^^^^^^^^
  File "/usr/local/Cellar/python@3.11/3.11.10/Frameworks/Python.framework/Versions/3.11/lib/python3.11/asyncio/runners.py", line 118, in run
    return self._loop.run_until_complete(task)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/Cellar/python@3.11/3.11.10/Frameworks/Python.framework/Versions/3.11/lib/python3.11/asyncio/base_events.py", line 654, in run_until_complete
    return future.result()
           ^^^^^^^^^^^^^^^
  File "/Users/mohammadmazraeh/Projects/autogen/python/packages/autogen-core/samples/distributed-group-chat/test_aoi.py", line 31, in main
    async for chunk in stream_output:
  File "/Users/mohammadmazraeh/Projects/autogen/python/packages/autogen-ext/src/autogen_ext/models/_openai/_openai_client.py", line 662, in create_stream
    choice.finish_reason
AttributeError: 'NoneType' object has no attribute 'finish_reason'
```

### What did you expect to happen?

I expected to get some reasonable response from `.create_stream` call as well. 

### How can we reproduce it (as minimally and precisely as possible)?

```python
import asyncio

from autogen_core.components.models._types import UserMessage
from autogen_ext.models._openai._openai_client import AzureOpenAIChatCompletionClient
from azure.identity import DefaultAzureCredential, get_bearer_token_provider

if __name__ == "__main__":

    async def main():
        config = {
            "model": "gpt-4o",
            "azure_endpoint": "https://xxxxxxx.openai.azure.com",
            "azure_deployment": "gpt-4o",
            "api_version": "2024-08-01-preview",
            "model_capabilities": {"vision": False, "function_calling": False, "json_output": False},
            "azure_ad_token_provider": get_bearer_token_provider(
                DefaultAzureCredential(), "https://cognitiveservices.azure.com/.default"
            ),
        }

        print(f"Creating client with config: {config}")
        client = AzureOpenAIChatCompletionClient(**config)

        messages = [UserMessage(content="Generate one short sentence on some topic!", source="system")]
        print("-----> Print output of .create call")
        single_output = await client.create(messages=messages)
        print("----->", single_output, "- * 50")

        print("-----> Print output of .create_stream call")
        stream_output = client.create_stream(messages=messages)
        async for chunk in stream_output:
            print(chunk)

    asyncio.run(main())

```

### AutoGen version

0.4.0.dev6

### Which package was this bug in

Extensions

### Model used

gpt-4o

### Python version

3.11.10

### Operating system

macOS Sequoia Version 15.1 (24B83)

### Any additional info you think would be helpful for fixing this bug

_No response_

### PR DESCRIPTION
**Pull request by MohMaz**
## Why are these changes needed?

I encountered this issue only in `AzureOpenAIChatCompletionClient`, adding some tolerance for empty `choices` ! Not a super clean fix ... just works to unblock updating streaming samples using `AzureOpenAIChatCompletionClient`

## Related issue number

closes #4213

## Checks

- [x] I've included any doc changes needed for https://microsoft.github.io/autogen/. See https://microsoft.github.io/autogen/docs/Contribute#documentation to build and test documentation locally.
- [x] I've added tests (if relevant) corresponding to the changes introduced in this PR.
- [x] I've made sure all auto checks have passed.

### CODE REVIEW

File: python/packages/autogen-ext/src/autogen_ext/models/_openai/_openai_client.py [hunk 1]
```diff
@@ -556,6 +556,7 @@ async def create_stream(
         json_output: Optional[bool] = None,
         extra_create_args: Mapping[str, Any] = {},
         cancellation_token: Optional[CancellationToken] = None,
+        max_empty_consequative_chunk_tolerance: Optional[int] = 0,
```
Comment (rysweet) [inline python/packages/autogen-ext/src/autogen_ext/models/_openai/_openai_client.py:559]: Should be "max_consecutive..."?

File: python/packages/autogen-ext/src/autogen_ext/models/_openai/_openai_client.py [hunk 2]
```diff
@@ -566,6 +567,7 @@ async def create_stream(
             json_output (Optional[bool], optional): If True, the output will be in JSON format. Defaults to None.
             extra_create_args (Mapping[str, Any], optional): Additional arguments for the creation process. Default to `{}`.
             cancellation_token (Optional[CancellationToken], optional): A token to cancel the operation. Defaults to None.
+            max_empty_consequative_chunk_tolerance (int, optional): The maximum number of consecutive empty chunks to tolerate before stopping the stream. Defaults to 3.
```
Comment (rysweet) [inline python/packages/autogen-ext/src/autogen_ext/models/_openai/_openai_client.py:570]: Looks like it defaults to zero?

File: python/packages/autogen-ext/src/autogen_ext/models/_openai/_openai_client.py [hunk 3]
```diff
@@ -566,6 +567,7 @@ async def create_stream(
             json_output (Optional[bool], optional): If True, the output will be in JSON format. Defaults to None.
             extra_create_args (Mapping[str, Any], optional): Additional arguments for the creation process. Default to `{}`.
             cancellation_token (Optional[CancellationToken], optional): A token to cancel the operation. Defaults to None.
+            max_consequative_empty_chunk_tolerance (int, optional): The maximum number of consecutive empty chunks to tolerate before stopping the stream. This seems to only be needed to set when using `AzureOpenAIChatCompletionClient`. Defaults to 0.
```
Comment (jackgerrits) [inline python/packages/autogen-ext/src/autogen_ext/models/_openai/_openai_client.py:570]: What does `none` mean?

Comment (jackgerrits) [inline python/packages/autogen-ext/src/autogen_ext/models/_openai/_openai_client.py:570]:
```suggestion
            max_consecutive_empty_chunk_tolerance (int, optional): The maximum number of consecutive empty chunks to tolerate before stopping the stream. This seems to only be needed to set when using `AzureOpenAIChatCompletionClient`. Defaults to 0.
```

File: python/packages/autogen-ext/src/autogen_ext/models/_openai/_openai_client.py [hunk 4]
```diff
@@ -636,13 +638,29 @@ async def create_stream(
...
         while True:
             try:
                 chunk_future = asyncio.ensure_future(anext(stream))
                 if cancellation_token is not None:
                     cancellation_token.link_future(chunk_future)
                 chunk = await chunk_future
 
+                # This is to address a bug in AzureOpenAIChatCompletionClient. OpenAIChatCompletionClient works fine.
+                #  https://github.com/microsoft/autogen/issues/4213
+                if len(chunk.choices) == 0:
```
Comment (jackgerrits) [inline python/packages/autogen-ext/src/autogen_ext/models/_openai/_openai_client.py:652]: Do the chunks with no choices have other data set on other fields of the delta?

File: python/packages/autogen-ext/src/autogen_ext/models/_openai/_openai_client.py (hunk 3 above)
Comment (jackgerrits) [inline python/packages/autogen-ext/src/autogen_ext/models/_openai/_openai_client.py:570]: It doesn't just stop the stream it errors out

File: python/packages/autogen-ext/src/autogen_ext/models/_openai/_openai_client.py (hunk 4 above)
Comment (MohMaz) [reply python/packages/autogen-ext/src/autogen_ext/models/_openai/_openai_client.py:652]: I can't see anything int the `chunk.__dict__`. In the consecutive chunks, the choices gets populated properly and has all the expected fields.
```
{'id': '', 'choices': [], 'created': 0, 'model': '', 'object': '', 'service_tier': None, 'system_fingerprint': None, 'usage': None}
```

File: python/packages/autogen-ext/src/autogen_ext/models/_openai/_openai_client.py (hunk 3 above)
Comment (MohMaz) [reply python/packages/autogen-ext/src/autogen_ext/models/_openai/_openai_client.py]: my bad, removed

File: python/packages/autogen-ext/src/autogen_ext/models/_openai/_openai_client.py [hunk 5]
```diff
@@ -636,13 +638,29 @@ async def create_stream(
...
+                    empty_chunk_count += 1
+                    if max_consecutive_empty_chunk_tolerance == 0:
+                        raise ValueError(
+                            "max_empty_ consecutive_chunk_tolerance is set to 0, increasing it might help!"
```
Comment (jackgerrits) [inline python/packages/autogen-ext/src/autogen_ext/models/_openai/_openai_client.py:656]:
```suggestion
                            "Consecutive empty chunks found. Change max_empty_consecutive_chunk_tolerance to increase empty chunk tolerance"
```

**Review by jackgerrits — APPROVED** [review]

**Comment by jspv** [comment]
Can we extend the fix to be an argument or property to the AzureOpenAIChatCompletionClient object rather than an argument to `create()` and `create_stream()` and have `create()` and `create_stream()` check the object?  This way Agents who call `create()` and `create_stream()` don't need to also be changed to implement this fix, the parameter can be set when instantiating the model client and then passed to the agent via `model_client=` when the Agent is instantiated.

**Pull request merged**